8Knot/_login.py

```diff
@@ -0,0 +1,216 @@
+            "client_id": os.environ.get("OAUTH_CLIENT_ID"),
+            "client_secret": os.environ.get("OAUTH_CLIENT_SECRET"),
+            "authorize_url": os.environ.get("OAUTH_AUTHORIZE_URL"),
+            "token_url": os.environ.get("OAUTH_TOKEN_URL"),
+            "redirect_uri": os.environ.get("OAUTH_REDIRECT_URI"),
+        }
+    }
+
+    # create flask-login object
+    login = LoginManager(server)
+    login.login_view = "/"
+
+    class User(UserMixin):
+        def __init__(self, id):
+            self.id = id
+
+    @login.user_loader
+    def load_user(id):
+        users_cache = redis.StrictRedis(
+            host="redis-users",
+            port=6379,
+            password=os.getenv("REDIS_PASSWORD", ""),
+        )
+        try:
+            users_cache.ping()
+        except redis.exceptions.ConnectionError:
+            logging.error("LOAD_USER: Could not connect to users-cache.")
+            return None
+
+        # return the JSON of a user that was set in the Redis instance
+        if users_cache.exists(id):
+            usn = json.loads(users_cache.get(id))["username"]
+            return User(id)
+        return None
+
+    @server.route("/logout/")
+    def logout():
+        users_cache = redis.StrictRedis(
+            host="redis-users",
+            port=6379,
+            password=os.getenv("REDIS_PASSWORD", ""),
+        )
+        try:
+            users_cache.ping()
+        except redis.exceptions.ConnectionError:
+            logging.error("LOGOUT: Could not connect to users-cache.")
+            return redirect("/")
+
+        if current_user.is_authenticated:
+            c_id = current_user.get_id()
+            users_cache.delete(c_id)
+            logout_user()
+            logging.warning(f"USER {c_id} LOGGED OUT")
+        else:
+            logging.warning("TRIED TO LOG OUT")
+        return redirect("/")
+
+    @server.route("/login/")
+    def oauth2_authorize():
+        users_cache = redis.StrictRedis(
+            host="redis-users",
+            port=6379,
+            password=os.getenv("REDIS_PASSWORD", ""),
+        )
+        try:
+            users_cache.ping()
+        except redis.exceptions.ConnectionError:
+            logging.error("LOGIN: Could not connect to users-cache.")
+            return redirect("/")
+
+        provider = os.environ.get("OAUTH_CLIENT_NAME")
+
+        if not current_user.is_anonymous:
+            return redirect("/")
+
+        provider_data = current_app.config["OAUTH2_PROVIDERS"].get(provider)
+        if provider_data is None:
+            abort(404)
+
+        # generate a random string for the state parameter
+        session["oauth2_state"] = secrets.token_urlsafe(16)
+
+        # create a query string with all the OAuth2 parameters
+        qs = urlencode(
+            {
+                "client_id": provider_data["client_id"],
+                # "redirect_uri": url_for("oauth2_callback", _external=True),
```

Comment:
Yeah it's a reminder for future context. 'redirect_uri' is a part of the standard but not implemented by Augur so I'm keeping its use noted in case Augur implements it in the future.

Reply:
Makes sense, good to know!